File: android/app/src/main/java/com/smartexpense/android/data/remote/RetrofitClient.java
package com.smartexpense.android.data.remote;

import android.content.Context;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitClient {
    // TODO: Update this to your local IP address for testing on physical device
    private static final String BASE_URL = "http://localhost:8080/";
    
    private static Retrofit retrofit = null;
    private static TokenManager tokenManager = null;

    public static void init(Context context) {
        if (tokenManager == null) {
            tokenManager = new TokenManager(context.getApplicationContext());
        }
    }

    public static TokenManager getTokenManager() {
        return tokenManager;
    }

    public static Retrofit getClient() {
        if (retrofit == null) {
            HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
            logging.setLevel(HttpLoggingInterceptor.Level.BODY);

            OkHttpClient client = new OkHttpClient.Builder()
                    .addInterceptor(logging)
                    .addInterceptor(new AuthInterceptor(tokenManager))
                    .build();

            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        return retrofit;
    }
}

File: android/app/src/main/kotlin/com/smartexpense/android/data/remote/RetrofitClient.kt
package com.smartexpense.android.data.remote

import android.content.Context
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitClient {
    // Update this to your local IP address for testing on physical device, or use 127.0.0.1 with adb reverse
    const val BASE_URL = "http://127.0.0.1:8080/"

    var retrofit: Retrofit? = null
        private set
    var tokenManager: TokenManager? = null
        private set

    fun init(context: Context) {
        if (tokenManager == null) {
            tokenManager = TokenManager(context.applicationContext)
        }
    }

    fun getClient(): Retrofit {
        if (retrofit == null) {
            val logging = HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BODY
            }

            val client = OkHttpClient.Builder()
                .addInterceptor(logging)
                .addInterceptor(AuthInterceptor(tokenManager!!))
                .build()

            retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        }
        return retrofit!!
    }
}
